docs: improve the component styling guide (#44234)
angular/angular: aio/content/guide/component-styles.md

<code-example path="component-styles/src/app/hero-app.component.ts" header="src/app/hero-app.component.ts"></code-example>

> ## Customizing component styles (or best practices)

Reviewer: I would make this either either just "Customizing component styles" or "Component styling best practices" (the parens in our draft was just me thinking that using the term "best practices" might be a good fit here)

> You should consider the styles of a component to be private implementation details for that component. When consuming a common component, you should not override the component's styles any more than you should access the private members of a TypeScript class. While Angular's style encapsulation prevents component styles from affecting other components, global styles affect all components on the page. This includes `::ng-deep`, which promotes a component style to a global style.

Reviewer 2: "while Angular's style encapsulation" may sound misleading... could you add a "default" to indicate that there are more options?

```diff
- While Angular's style encapsulation prevents ...
+ While Angular's default style encapsulation prevents ...
```
or you could be a bit more precise:
```diff
- While Angular's style encapsulation prevents ...
+ While Angular's default view encapsulation strategy prevents ...
```

> You should consider the styles of a component to be private implementation details for that component. When consuming a common component, you should not override the component's styles any more than you should access the private members of a TypeScript class. While Angular's default style encapsulation prevents component styles from affecting other components, global styles affect all components on the page. This includes `::ng-deep`, which promotes a component style to a global style.

Reviewer: We might want to add an infobox above this paragraph with something like 
```markdown
"See [View Encapsulation](...) for information on how Angular scopes styles to specific components."
```

### Authoring a component to support customization

As component author, you can explicitly design a component to accept customization in one of four different ways.

> #### 1. Use CSS Custom Properties (Recommended)

Reviewer: ```suggestion
#### 1. Use CSS Custom Properties (recommended)
```
nit: capitalization

> You can define a supported customization API for your component by defining its styles with CSS Custom Properties, alternatively known as CSS Variables. Component library users consume this API by defining values for these properties, customizing the final appearance of the component on the rendered page.

Reviewer 2: ```suggestion
You can define a supported customization API for your component by defining its styles with CSS Custom Properties, also known as CSS Variables. Component library users consume this API by defining values for these properties, customizing the final appearance of the component on the rendered page.
```
just a prederence.... alternatively doesn't sound very natural there to me :sweat_smile:

> You can define a supported customization API for your component by defining its styles with CSS Custom Properties, alternatively known as CSS Variables. Component library users consume this API by defining values for these properties, customizing the final appearance of the component on the rendered page.

Reviewer: ```suggestion
You can define a supported customization API for your component by defining its styles with CSS Custom Properties, alternatively known as CSS Variables. Anyone using your component can consume this API by defining values for these properties, customizing the final appearance of the component on the rendered page.
```
realized that "component library users" is overly specific since it can be _any_ component

> While this requires defining a custom property for each customization point, it creates a clear API contract that works in all style encapsulation modes.

Reviewer 2: ```suggestion
While this requires defining a custom property for each customization point, it creates a clear API contract that works in all view encapsulation modes.
```

#### 2. Declare global CSS with @mixin

> While Angular's emulated style encapsulation prevents styles from escaping a component, it does not prevent global CSS from affecting the entire page. While component consumers should avoid directly overwriting the CSS internals of a component, you can offer a supported customization API via a CSS preprocessor like Sass.

Reviewer 2: ```suggestion
While Angular's emulated view encapsulation prevents styles from escaping a component, it does not prevent global CSS from affecting the entire page. While component consumers should avoid directly overwriting the CSS internals of a component, you can offer a supported customization API via a CSS preprocessor like Sass.
```

For example, a component may offer one or more supported mixins to customize various aspects of the component's appearance. While this approach uses global styles in it’s implementation, it allows the component author to keep the mixins up to date with changes to the component's private DOM structure and CSS classes.

#### 3. Customize with CSS ::part

> If your component uses [Shadow DOM](https://developer.mozilla.org/docs/Web/Web_Components/Using_shadow_DOM), you can apply the `part` attribute to specify elements in your component's template. This allows consumers of the component to author arbitrary styles targeting those specific elements with [the `::part` pseudo-element](https://developer.mozilla.org/docs/Web/CSS/::part).

Reviewer 2: ```suggestion
If your component uses the ShadowDom view encapsulation strategy, you can apply the `part` attribute to specify elements in your component's template. This allows consumers of the component to author arbitrary styles targeting those specific elements with [the `::part` pseudo-element](https://developer.mozilla.org/docs/Web/CSS/::part).
```
the way you put it sounds like this works when the component uses the shadow dom whether is uses that view encapsulation mode or not? which I guess it's wrong :thinking: :shrug: 
(never tried using the shadow dom in the emulated mode... not too sure what the result would be :sweat_smile: .... well it would be terribly hacky that's for sure :sweat_smile:)
